### ISSUE TITLE
"Latest-tag" should be upgrade to "non-fixed-tag check"

### ISSUE THREAD

**Issue by DarthSlider** [digitalocean/clusterlint#63]
We just found that one of our users used "nightly" tag in his manifest so I think that it should check for any non-fixed versions of images.

**Comment by adamwg** [comment]
Hm, this is an interesting suggestion. How would we reliably distinguish between "fixed" and "non-fixed" tags? `latest` is a special case since it's the default in the docker client if no tag is specified when building an image. While some projects use semantic versions or similar for images, there are also many that don't (e.g., `debian:stretch`, `alpine:3`). I'd be hesitant to have clusterlint throw a warning for a non-semver image tag.

**Comment by timoreimann** [comment]
I wonder if we could solve this by making certain checks more configurable (beyond turning checks entirely on or off). I don't think it's something clusterlint fundamentally supports today though, so it might additional effort (in terms of design and implementation).

Curious what others think.

**Comment by adamwg** [comment]
I was talking with @nathanjsweet about this on Friday and he had a suggestion I liked. He suggested implementing a much stricter check, requiring that all images are specified by SHA rather than by tag - if you want to make sure your images don't change that is the best practice. I think we would want to put this check in a different group (maybe the `security` group, or maybe a new `strict` group or something like that) since specifying images by SHA is not that common despite being a good idea.

**Comment by timoreimann** [comment]
@varshavaradarajan I agree that what constitutes a "bad" tag may vary, presumably depending on how strict a specific organization is about certain tags. My thinking was that we could offer a generic mechanism for clusterlint users to specify what these bad tags are, e.g.,

```shell
clusterlint -g basic --check-options-latest-tag bad-tags='latest,nightly,0.*'
```

That way, we (as clusterlint maintainers) do not need to bother with catching up on what the latest bad tags are. It'd up to the user to decide and specify per their individual needs.

This may also meet the needs of users who explicitly expect tags to be overridable. For instance, `debian:stable` is guaranteed to always point at the latest stable, and `debian:stretch` will always include the latest security updates for stretch; both is by (Docker Hub) design.

My suggestion would certainly involve more effort since we do not have a mechanism in place today to pass through arbitrary check arguments. The SHA-based approach also sounds interesting but seems to address a slightly different use case (i.e., super strict mode).

Maybe @DarthSlider has some additional thoughts on which approach might best suit their needs.

**Comment by nathanjsweet** [comment]
I opened an [image check PR](https://github.com/digitalocean/clusterlint/issues/64 that I thought might address these issues.

**Comment by DarthSlider** [comment]
According to our local policies, every word-only tags are considered as bad (except stretch I think).
It might not be the best practice ever, but I think it's safe enough.
So, I assume that clusterlint could just give a warning for every tag without numbers in it, or, for more strict check for every tag without X.Y in it.
